yt/fields/magnetic_field.py

```diff
@@ -119,59 +120,121 @@ def _magnetic_field_toroidal(field, data):
-            r = data["index", "r"]
-            z = data["index", "z"]
-            d = np.sqrt(r * r + z * z)
-            rax = axis_names.find("r")
-            zax = axis_names.find("z")
-            return (data[ftype, "magnetic_field_r"] - bm[rax]) * (r / d) + (
-                data[ftype, "magnetic_field_z"] - bm[zax]
-            ) * (z / d)
-
-        def _magnetic_field_toroidal(field, data):
+            rax = axis_names.index("r")
+            zax = axis_names.index("z")
+
+            return np.sqrt(
+                (data[ftype, "magnetic_field_r"] - bm[rax]) ** 2
+                + (data[ftype, "magnetic_field_z"] - bm[zax]) ** 2
+            )
+
+        def _magnetic_field_toroidal_magnitude(field, data):
             ax = axis_names.find("theta")
             bm = data.get_field_parameter("bulk_magnetic_field")
             return data[ftype, "magnetic_field_theta"] - bm[ax]
 
     elif registry.ds.geometry == "spherical":
 
-        def _magnetic_field_poloidal(field, data):
-            ax = axis_names.find("theta")
+        def _magnetic_field_poloidal_magnitude(field, data):
             bm = data.get_field_parameter("bulk_magnetic_field")
-            return data[ftype, "magnetic_field_theta"] - bm[ax]
+            rax = axis_names.index("r")
+            tax = axis_names.index("theta")
+
+            return np.sqrt(
+                (data[ftype, "magnetic_field_r"] - bm[rax]) ** 2
+                + (data[ftype, "magnetic_field_theta"] - bm[tax]) ** 2
+            )
 
-        def _magnetic_field_toroidal(field, data):
+        def _magnetic_field_toroidal_magnitude(field, data):
             ax = axis_names.find("phi")
             bm = data.get_field_parameter("bulk_magnetic_field")
             return data[ftype, "magnetic_field_phi"] - bm[ax]
 
-    if _magnetic_field_toroidal is not None:
+    if _magnetic_field_poloidal_magnitude is not None:
         registry.add_field(
-            (ftype, "magnetic_field_toroidal"),
+            (ftype, "magnetic_field_poloidal_magnitude"),
             sampling_type="local",
-            function=_magnetic_field_toroidal,
+            function=_magnetic_field_poloidal_magnitude,
             units=u,
             validators=[
                 ValidateParameter("normal"),
                 ValidateParameter("bulk_magnetic_field"),
             ],
         )
-    if _magnetic_field_poloidal is not None:
+
+    if _magnetic_field_toroidal_magnitude is not None:
         registry.add_field(
-            (ftype, "magnetic_field_poloidal"),
+            (ftype, "magnetic_field_toroidal_magnitude"),
             sampling_type="local",
-            function=_magnetic_field_poloidal,
+            function=_magnetic_field_toroidal_magnitude,
             units=u,
             validators=[
                 ValidateParameter("normal"),
                 ValidateParameter("bulk_magnetic_field"),
             ],
         )
 
+    if registry.ds.geometry == "cartesian":
```

Comment:
I guess I am a little confused as to why in the `"cartesian"` geometry the fields corresponding to the coordinate directions (r, theta, phi) need to be specified explicitly. If we set up `"velocity"` as a vector field in cartesian, we get the cartesian, cylindrical, and spherical coordinate component fields all at once. Does that not happen for the magnetic field for some reason?

Reply:
You're right, we already get `"magnetic_field_spherical_phi"` and `"magnetic_field_spherical_theta"` for free. I'll simplify this, thank you very much for pointing it out !

On a general note, most of what I'm doing here could be done in `vector_operations.py` and be usable for both magnetic and velocity fields, and I will very likely make some deduplications in the future. Here I've been trying to get straight to a usable state.